### PR TITLE
Add bee.watch([range], [onchange])

### PR DESCRIPTION
**Pull request by LuKks**

### CODE REVIEW

File: index.js [hunk 1]
```diff
@@ -835,6 +856,89 @@ class Batch {
...
+      this.destroy()
+    } finally {
+      this.running = false
+    }
+  }
+
+  async _run () {
+    this.running = true
+
+    // + I think it could repeat a last call even after bee is closed, should trigger it with a test first
```
Comment (HDegroote) [inline index.js:894]: Yeah, indeed, and there's another race condition too. Not sure if we need to care about these, but they could lead to some confusion:
- A watcher can be unwatched/destroyed while an onAppend is being processed. So in that case too you would observe an event after closing. I believe the solution there is to make this.running a promise or null, and adding
`if (this.running) await this.running` after setting this.watching to false

File: README.md [hunk 2]
```diff
@@ -230,6 +230,27 @@ returned, only the diff.
...
+#### `watcher = db.watch([range], [onchange])`
+
+Listens to changes that are on the optional `range`.
+
+`range` options are the same as `createReadStream` except for reverse.
+
+`watcher.watch()`\
+Starts watching for changes. By default it already does it. This is only useful after you `unwatch()`.
+
+`watcher.unwatch()`\
```
Comment (HDegroote) [inline README.md:242]: Do you have a usecase in mind for unwatching? On first sight, it seems like it could also be done by destroying the watcher and creating a new watcher later

File: index.js (hunk 1 above)
Comment (HDegroote) [reply index.js:894]: Ah, just understood your logic with storing this.stream as instance variabele. That also solves the race condition I described, and probably in a cleaner way (it won't be forced to wait until all diffs are yielded)

File: index.js [hunk 3]
```diff
@@ -835,6 +856,89 @@ class Batch {
...
+    // if (this.destroyed) return
+
+    const snapshot = this.bee.snapshot()
+    this.stream = snapshot.createDiffStream(this.latestDiff, this.range)
+
+    try {
+      for await (const data of this.stream) { // eslint-disable-line
+        this.emit('change', snapshot.version, this.latestDiff)
```
Comment (HDegroote) [inline index.js:902]: Any particular reason to not emit the actual diff? That's what I would be most interested in most of the time

Comment (LuKks) [reply index.js:902]: Yes, but if I give you only the first diff and skip over all others (due `break`ing the loop) then the diff is not useful

Comment (LuKks) [reply index.js:902]: That's why I wanted a `opts.includeDiff` but Maf spoke haha

File: README.md (hunk 2 above)
Comment (LuKks) [reply README.md:242]: I took the idea from `udx` actually, and it was easy to do the change

File: index.js (hunk 3 above)
Comment (HDegroote) [reply index.js:902]: Ah, should have read that thread with more attention. Yeah, sounds reasonable

File: README.md (hunk 2 above)
Comment (LuKks) [reply README.md:242]: But I agree it could be removed on this case

Comment (LuKks) [reply README.md:242]: Also I agree even more because `unwatch()` is what introduced `watching` property, so maybe it's good if we just remove unwatch lol more simple

File: index.js [hunk 4]
```diff
@@ -382,6 +384,25 @@ class Hyperbee {
...
 
+  watch (range, onchange) {
+    if (typeof range === 'function') {
+      onchange = range
+      range = undefined
+    }
+
+    const watcher = new Watcher(this, range)
+
+    this.feed.once('close', () => {
```
Comment (HDegroote) [inline index.js:395]: Not sure how these event handlers work, but what happens to this event if the watcher closes, but the feed does not? Will the event still exist? If so, that's a (tiny) memory leak

Comment (LuKks) [reply index.js:395]: Sounds like it is, I will add a test creating hundreds of watchers

File: index.js (hunk 1 above)
Comment (LuKks) [reply index.js]: I can reproduce it by manually adding a delay before the diff stream loop (i.e. simulates network)

I couldn't reproduce it on a test but it's obviously possible. We don't want to trigger any `'change'` event or any error if the user already manually did `destroy()` so I think we can safely return right there

File: index.js [hunk 5]
```diff
@@ -835,6 +871,68 @@ class Batch {
...
+    this.destroyed = false
+
+    this.range = range
+    this.latestDiff = this.bee.version
+    this.stream = null
+
+    this._onappend = this._onappend.bind(this)
+    this._onappend = debounceify(this._onappend)
+
+    // this.on('error', noop)
```
Comment (LuKks) [inline index.js:891]: Leaving this to @mafintosh to decide if it should ignore errors by default, otherwise watcher can throw errors due `emit('error', err)` but only unexpected ones

File: index.js [hunk 6]
```diff
@@ -835,6 +871,68 @@ class Batch {
...
+      this.destroy()
+    } finally {
+      this.running = false
+    }
+  }
+
+  async _run () {
+    if (this.destroyed) return
+
+    const snapshot = this.bee.snapshot()
```
Comment (LuKks) [inline index.js:910]: @HDegroote is this finally needed vs just storing the version?

File: index.js (hunk 1 above)
Comment (LuKks) [reply index.js]: I think this could be solved if we could cancel the next debounce call if any, but it's much simpler to just check if destroyed and stop the execution as we're doing now

Comment (HDegroote) [reply index.js]: I didn't think it was possible to cancel something which has been scheduled on the event loop. Is it?

File: index.js (hunk 6 above)
Comment (HDegroote) [reply index.js:910]: (tl;dr using snapshots avoids headaches)

Yes, I believe it is. Without a snapshot the following can happen:
- Suppose current version is 10, lastDiff 5
- Create a diffStream compared to 5 (asuming current version is 10)
- await the first entry of the stream
- Before the first entry resolves, a new element is added to the bee (so version 11)
- The first entry of the stream resolves, and the diffStream snaps from version 5->11

Without the snapshot it becomes pretty difficult to reason about. It's even tricky to figure out between which versions you have taken the diffStream (you can no longer do `    this.latestDiff = snapshot.version`. Doing `    this.latestDiff = this.bee.version` at the end could be wrong (bee can have updated), and doing it at the start could also be wrong as already illustrated.

From what I remember when investigating this, there's actually no safe way to know between which versions you have taken the diffStream of a non-snapshot bee, because it's possible for new entries to be added to the bee between the time the diffStream snaps and the time it returns control to the consumer for the first time.

It's a bit tricky to trigger in tests because the race conditions depend on the order on the event loop, which to some extend depends on CPU speed, but on my machine this triggers it:

First get rid of the snapshot:
`    const snapshot = this.bee // .snapshot() `

```
test.solo('multiple changes', async function (t) {
  t.plan(2)

  const db = create()

  const watcher = db.watch()
  t.teardown(() => watcher.destroy())

  watcher.on('change', function (leftVersion, rightVersion) {
    console.log(leftVersion, rightVersion)
  })

  await Promise.all([db.put('/a'), db.put('/b'), db.put('/c'), db.put('/d'), db.put('/e')])
  await sleep(500)
})
```
output:
```
3 1
4 3
6 5
```

so no trigger for 4->5

File: index.js [hunk 7]
```diff
@@ -290,6 +292,10 @@ class Hyperbee {
     this._checkout = opts.checkout || 0
     this._ready = opts._ready || null
 
+    this._watchers = new Set()
+    this._onappend = this._onappend.bind(this)
```
Comment (mafintosh) [inline index.js:296]: this_onappendbound

File: README.md [hunk 8]
```diff
@@ -230,6 +230,21 @@ returned, only the diff.
...
+#### `watcher = db.watch([range], [onchange])`
+
+Listens to changes that are on the optional `range`.
+
+`range` options are the same as `createReadStream` except for reverse.
+
+`watcher.destroy()`\
+Closes the watcher.
+
+`watcher.on('change', (leftVersion, rightVersion) => {})`\
```
Comment (mafintosh) [inline README.md:242]: newVersion, oldVersion I'd say here

File: index.js (hunk 6 above)
Comment (mafintosh) [reply index.js:910]: @LuKks we should add an option to readstream, `head` thats the length at which it starts, then its trivial to not use a snap

Comment (mafintosh) [reply index.js:910]: but we can do that in a follow up pr

File: index.js [hunk 9]
```diff
@@ -835,6 +871,68 @@ class Batch {
...
+    const snapshot = this.bee.snapshot()
+    this.stream = snapshot.createDiffStream(this.latestDiff, this.range)
+
+    try {
+      for await (const data of this.stream) { // eslint-disable-line
+        this.emit('change', snapshot.version, this.latestDiff)
+        break
+      }
+    } catch (err) {
+      if (this.destroyed && err.message === 'Stream was destroyed') return
```
Comment (mafintosh) [inline index.js:919]: if (this.destroyed) is better

File: index.js [hunk 10]
```diff
@@ -835,6 +871,68 @@ class Batch {
...
+    }
+
+    this.latestDiff = snapshot.version
+  }
+
+  destroy () {
+    if (this.destroyed) return
+    this.destroyed = true
+
+    if (this.stream) this.stream.destroy()
```
Comment (mafintosh) [inline index.js:930]: wait for the stream to emit close here

File: index.js [hunk 11]
```diff
@@ -835,6 +871,68 @@ class Batch {
...
+    try {
+      for await (const data of this.stream) { // eslint-disable-line
+        this.emit('change', snapshot.version, this.latestDiff)
+        break
+      }
+    } catch (err) {
+      if (this.destroyed && err.message === 'Stream was destroyed') return
+      throw err
+    }
+
```
Comment (mafintosh) [inline index.js:922]: `this.stream = null` here

File: index.js [hunk 12]
```diff
@@ -835,6 +871,68 @@ class Batch {
...
+    // this.on('error', noop)
+  }
+
+  async _onappend () {
+    this.running = true
+
+    try {
+      await this._run()
+    } catch (err) {
+      this.emit('error', err)
```
Comment (mafintosh) [inline index.js:900]: i would just emit warning here instead of error (solves your question above).

File: index.js [hunk 13]
```diff
@@ -835,6 +871,68 @@ class Batch {
...
+
+    this.running = false
+    this.destroyed = false
+
+    this.range = range
+    this.latestDiff = this.bee.version
+    this.stream = null
+
+    this._onappend = this._onappend.bind(this)
+    this._onappend = debounceify(this._onappend)
```
Comment (mafintosh) [inline index.js:889]: onappendBound like above and do both of these in one go

File: index.js [hunk 14]
```diff
@@ -835,6 +871,68 @@ class Batch {
...
+    super()
+
+    this.bee = bee
+    this.core = bee.core
+
+    this.running = false
+    this.destroyed = false
+
+    this.range = range
+    this.latestDiff = this.bee.version
```
Comment (mafintosh) [inline index.js:885]: you should set this in _run for the first run, ie the "ready" step. otherwise the bee has to be ready'ed first, which should not be a requirement

Comment (mafintosh) [reply index.js:885]: ie

```js
_run () {
  if (first) {
    await this.bee.ready()
    this.latestDiff = this.bee.version
    return
  }
  ...
}
```

And then trigger the debouncer once from the constructor manually with a safetyCatch .catch

File: index.js (hunk 10 above)
Comment (LuKks) [reply index.js]: Should I use `destroying`?
https://github.com/holepunchto/hyperbee/blob/bb99190279b2263bef26848a3b318f074b543f8a/index.js#L933
With `destroyed` also works, but sounds like if `destroying` is `true` then no need to `destroy()` the stream

Another version, this sounds even more precise but it could be unnecessary:
```js
if (this.stream && !this.stream.destroyed) {
    if (!this.stream.destroying) this.stream.destroy()
    this.stream.once('close', emitClose)
  } else {
  setImmediate(emitClose)
}
```

File: index.js (hunk 12 above)
Comment (LuKks) [reply index.js:899]: Wouldn't be better to just noop the `error` event? As it's a more common event, also notice that the error should only happen if user misuses the watcher

For example, no watch related, this throws `SESSION_CLOSED`:
```js
const put1 = db.put('asd')
await db.close()
await put1
```
Because let's say user is misusing `put()`

Example of watch throwing an error:
https://github.com/holepunchto/hyperbee/blob/bb99190279b2263bef26848a3b318f074b543f8a/test/watch.js#L216

File: index.js (hunk 14 above)
Comment (LuKks) [reply index.js:885]: I was thinking about the ready step at the beginning then I did this this test to prove that is not needed:
https://github.com/holepunchto/hyperbee/blob/ee5d3f409f47d127a410cb2599b9b97859f2d25d/test/watch.js#L87
But you're right, and also just realized the test is wrong

Comment (LuKks) [reply index.js]: Okey, this is the new test that confirms that it should be working good now:
https://github.com/holepunchto/hyperbee/blob/1bec4e71bfa373623afc3758cb12bc769ffe73fe/test/watch.js#L130
`oldVersion` was resulting in `1` before, but now `3` which sounds correct

File: index.js (hunk 1 above)
Comment (LuKks) [reply index.js]: I meant the `debounceify` debounced last call, but overkill

File: README.md [hunk 15]
```diff
@@ -230,6 +230,25 @@ returned, only the diff.
...
+
+`watcher.on('change', (newVersion, oldVersion) => {})`\
+Emitted after a feed change.
+
+`watcher.on('error', onerror)`\
+Critical and unexpected errors will be thrown, but watcher is normally graceful.\
+Also, watcher auto closes on errors.
+
+`watcher.on('close', onclose)`\
+Emitted after the watcher is closed.
```
Comment (mafintosh) [inline README.md:250]: why the escaped new lines?

File: test/helpers/index.js [hunk 16]
```diff
@@ -70,10 +78,41 @@ async function toString (tree) {
...
 function createCore () {
   return new Hypercore(require('random-access-memory'))
 }
+
+function createTmpDir (t) {
+  const tmpdir = path.join(os.tmpdir(), 'hyperbee-test-')
+  const dir = fs.mkdtempSync(tmpdir)
+  t.teardown(() => fs.rmSync(dir, { recursive: true, force: true }))
+  return dir
+}
```
Comment (mafintosh) [inline test/helpers/index.js:110]: remove this functino

File: test/watch.js [hunk 17]
```diff
@@ -0,0 +1,318 @@
...
+  const watcher = db.watch({ gte: '14' })
+  t.teardown(() => watcher.destroy())
+
+  const onchangefail = () => t.fail('should not trigger changes')
+  const onchangepass = () => t.pass('change')
+
+  watcher.on('change', onchangefail)
+  await db.put('13')
+  await eventFlush()
+  await sleep(500)
```
Comment (mafintosh) [inline test/watch.js:58]: why is a sleep needed?

File: test/watch.js [hunk 18]
```diff
@@ -0,0 +1,318 @@
...
+
+  db.watch(function () {
+    t.fail('should not trigger changes')
+  })
+
+  await db.ready()
+  t.is(db.version, 3)
+
+  await eventFlush()
+  await sleep(500)
```
Comment (mafintosh) [inline test/watch.js:108]: why is a sleep needed?

File: test/watch.js [hunk 19]
```diff
@@ -0,0 +1,318 @@
...
+
+  await db.close()
+
+  t.pass()
```
Comment (mafintosh) [inline test/watch.js:112]: why is this needed?

File: test/watch.js [hunk 20]
```diff
@@ -0,0 +1,318 @@
...
+    t.is(newVersion, 4)
+    t.is(oldVersion, 3)
+  })
+
+  await db.put('/c')
+  await eventFlush()
+
+  await db.close()
+
+  t.pass()
```
Comment (mafintosh) [inline test/watch.js:138]: why is this needed?

File: test/watch.js [hunk 21]
```diff
@@ -0,0 +1,318 @@
...
+  const db = create()
+  t.is(db.version, 1)
+
+  db.watch(function (newVersion, oldVersion) {
+    t.is(newVersion, 4)
+    t.is(oldVersion, 3)
+  })
+
+  await db.put('/c')
+  await eventFlush()
```
Comment (mafintosh) [inline test/watch.js:134]: why is it event flushing here? needs a comment

Comment (mafintosh) [reply test/watch.js:134]: same for the other ones above

File: test/watch.js [hunk 22]
```diff
@@ -0,0 +1,318 @@
...
+  t.absent(watcher.destroyed)
+  watcher.destroy()
+  t.ok(watcher.destroyed)
+
+  await db.put('/a')
+
+  await eventFlush()
+  await sleep(500)
+
+  t.pass()
```
Comment (mafintosh) [inline test/watch.js:166]: same as above

File: test/watch.js [hunk 23]
```diff
@@ -0,0 +1,318 @@
...
+    })
+
+    t.absent(watcher.destroyed)
+    watcher.destroy()
+    t.ok(watcher.destroyed)
+  })
+
+  await db.put('/a')
+
+  t.pass()
```
Comment (mafintosh) [inline test/watch.js:188]: same as above

File: test/watch.js [hunk 24]
```diff
@@ -0,0 +1,318 @@
...
+
+  watcher.on('error', function (err) {
+    t.fail('should not have given error: ' + err)
+  })
+
+  await put
+  await eventFlush()
+  await sleep(500)
+
+  t.pass()
```
Comment (mafintosh) [inline test/watch.js:236]: same as above

File: test/watch.js [hunk 25]
```diff
@@ -0,0 +1,318 @@
...
+    db.core.close()
+  })
+
+  watcher.on('error', function (err) {
+    t.is(err.code, 'SESSION_CLOSED')
+  })
+
+  await db.put('/b')
+
+  t.pass()
```
Comment (mafintosh) [inline test/watch.js:264]: same as above

File: test/watch.js [hunk 26]
```diff
@@ -0,0 +1,318 @@
...
+    await eventFlush()
+
+    if (!changed) {
+      t.fail('should have changed')
+    }
+
+    watcher.destroy()
+  }
+
+  t.pass()
```
Comment (mafintosh) [inline test/watch.js:317]: same as above

File: README.md (hunk 15 above)
Comment (LuKks) [reply README.md:251]: On this case, with escape vs without vs with break line:

image

=>

image

Comment (LuKks) [reply README.md:251]: I don't like the extra space in between that a normal break line does, so that's why I escape them, but I can remove the escapes if you like

File: test/watch.js (hunk 21 above)
Comment (LuKks) [reply test/watch.js:130]: That could be an issue as watcher is not graceful when it's destroyed

First, when doing a `db.put` it needs an event flush for watch to trigger a `change` event

With that in mind, when you do `await db.close()` all watchers gets instantly destroyed, so without waiting an event flush there will be no `change` event

Comment (LuKks) [reply test/watch.js:130]: So the question is, should we support triggering a change event on this case:
```js
db.watch(function (newVersion, oldVersion) {
  // change?
})

await db.put('/c')

await db.close()
```

Sounds like it should

Comment (LuKks) [reply test/watch.js:130]: Hm even if it looks like it should, adding a simple event flush on the close would not be enough for a diff stream with networking involved. We could make watcher destroy async but making `bee.close` to wait doesn't sound good

Comment (LuKks) [reply test/watch.js:130]: Maybe we can special case it only for the writter?
